src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -0,0 +1,629 @@
+        let filteredInterpreters = [...interpreters];
+
+        // Then go through all of the kernels and generate their metadata
+        const kernelMetadata = kernelSpecs.map((k) => {
+            // Find the interpreter that matches. If we find one, we want to use
+            // this to start the kernel.
+            const matchingInterpreters = this.findMatchingInterpreters(k, interpreters);
+            if (matchingInterpreters && matchingInterpreters.length) {
+                const result: PythonKernelConnectionMetadata = {
+                    kind: 'startUsingPythonInterpreter',
+                    kernelSpec: k,
+                    interpreter: matchingInterpreters[0],
+                    id: getKernelId(k, matchingInterpreters[0])
+                };
+
+                // If interpreters were found, remove them from the interpreter list we'll eventually
+                // return as interpreter only items
+                filteredInterpreters = filteredInterpreters.filter((i) => !matchingInterpreters.includes(i));
+
+                // Return our metadata that uses an interpreter to start
+                return result;
+            } else {
+                // No interpreter found. If python, use the active interpreter anyway
+                const result: KernelSpecConnectionMetadata = {
+                    kind: 'startUsingKernelSpec',
+                    kernelSpec: k,
+                    interpreter: k.language === PYTHON_LANGUAGE ? activeInterpreter : undefined,
+                    id: getKernelId(k, activeInterpreter)
+                };
+                return result;
+            }
+        });
+
+        // Combine the two into our list
+        const results = [
+            ...kernelMetadata,
+            ...filteredInterpreters.map((i) => {
+                // Update spec to have a default spec file
+                const spec = createIntepreterKernelSpec(i, rootSpecPath);
+                const result: PythonKernelConnectionMetadata = {
+                    kind: 'startUsingPythonInterpreter',
+                    kernelSpec: spec,
+                    interpreter: i,
+                    id: getKernelId(spec, i)
+                };
+                return result;
+            })
+        ];
+
+        // Sort them so that the active interpreter comes first (if we have one for it).
+        // This allows searches to prioritize this kernel first. If you sort for
+        // a UI do it after this function is called.
+        return results.sort((a, b) => {
+            if (a.kernelSpec?.display_name === b.kernelSpec?.display_name) {
+                return 0;
+            } else if (
+                a.interpreter?.path === activeInterpreter?.path &&
+                a.kernelSpec?.display_name === activeInterpreter?.displayName
+            ) {
+                return -1;
+            } else {
+                return 1;
+            }
+        });
+    }
+
+    private findMatchingInterpreters(
```

Comment:
This looks like another spot where (from looking at customer logs) adding logging in the new extra kernel logging option might be helpful.

Reply:
I'll add some logging here.